CI: Rework how docs are built

errordeveloper opened the pull request:
The aim is to make feedback on documentation PR faster, this probably the first phase of changes, we might want to add preview site deployments etc.

Fixes #10555

This runs very fast (under a minute) on GitHub Actions: https://github.com/errordeveloper/cilium/runs/505646021?check_suite_focus=true

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10572)
<!-- Reviewable:end -->

errordeveloper committed:
docs: Simplify the build

- Use docker image build cache better
- Move `make check-build` logic into a script and make it more readable
- Move `make cmdref` logic into a script

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
make: Simplify docs targets

- Remove `make check-docs` that was primarily intended to fix
  dependency issues (see 971c514052b8c8e8b056c545e64cc67db8a51333),
  it's no longer necessary as the documentation build assumes
  containerised build
- Keep `make render-docs`, as it's commonly used

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
CI: Add GitHub Actions Workflow

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,14 @@
+        runs-on: ubuntu-18.04
+        steps:
+            - uses: actions/checkout@v1
+            - uses: docker://errordeveloper/cilium-docs:latest
```
I pushed my own image for now, let's discuss where this one should live and if we need to tag and push it automatically.

joestringer replied:
We already have https://hub.docker.com/r/cilium/docs-builder, presumably we will just push a new version of that image?

errordeveloper replied:
For sure, I just need push access to begin with :)

I'd be happy to automate pushing the image, but perhaps best to do it in another PR?

joestringer replied:
Usually I think they just pull from repo updates and build on the remote infrastructure, is there something specific we need to automate here?

joestringer replied:
Ah, maybe we only do this for the main cilium image so if the docs base target doesn't change often then we don't need to keep building a new image.

errordeveloper replied:
Yes, this one shouldn't change very often.

errordeveloper commented on jenkinsfiles/docs.Jenkinsfile:
```diff
@@ -44,7 +44,7 @@ pipeline {
                 Status("PENDING", "${env.JOB_NAME}")
                 checkout scm
-                sh "make test-docs"
+                sh "make -C Documentation html"
```
Should we actually keep this job or disable it to reduce load Jenkins (if nothing else)?

joestringer replied:
I'm all for disabling if this is handled by another CI target and that target is working as intended.

aanm replied:
+1 we need it in the CI mostly to detect misspelled words

errordeveloper committed:
docs: Add note on linkchecker

Leave the code commented out as many links are broken right now,
and it's slow.

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented on Documentation/check-build.sh:
```diff
@@ -0,0 +1,81 @@
+    fi
+fi
+
+# TODO: Fix broken links and re-enable this
```
I attempted fixing some of the broken links (https://github.com/cilium/cilium/pull/10576), but it looks like Sphinx' link checker will ways fail on some of our links (e.g. things it spots in our docs like `curl https://exampe.com/$(uname)`), it also seems to fail to find anchors also.

Perhaps we should just drop this, or explore ways of controlling the behaviour?

joestringer replied:
Just because we don't run the target doesn't mean links stop being dead ;)

Since we don't regularly run this check, some links have rotten. It'd be nice to be able to automate this as well, but I wouldn't make fixing this target a precondition to merging this PR.

errordeveloper replied:
Sure, we should keep up with links, but we have some links that will keep on failing because of how this link checker works (see https://github.com/cilium/cilium/pull/10576#issuecomment-599504742).

aanm replied:
I think I've ran this somewhere manually in the past. The main issue with it was that is extremely slow since it's literally checking if the URL exists or not.

errordeveloper replied:
Yes, and it appears to me that does those checks one-by-one in a single Python thread :)

joestringer commented on Documentation/check-build.sh:
```diff
@@ -0,0 +1,81 @@
+		describe_spelling_errors
+        hint_about_wordlist_update
+		status_ok=1
+	fi
```
nit:
```suggestion
        describe_spelling_errors
        hint_about_wordlist_update
	status_ok=1
    fi
```

(I think your editor introduced some tabs in here unintentionally)

errordeveloper replied:
yeah... I tend to mix vim/vscode for shell scripts and some other files, will fix this :)

joestringer commented on Documentation/check-build.sh:
```diff
@@ -0,0 +1,81 @@
+}
+
+echo "Validating documentation (syntax, spelling)..."
+if build_with_spellchecker ; then
```
I *think* that previously if the spellchecker build failed (non-zero exit code), we would exit immediately. With this script, I suspect we will just skip this conditional block and proceed to attempt to build again without the spellchecker target, so you'll end up running sphinx twice in such cases?

errordeveloper replied:
Yes, that was intentional. The aim was to catch our fatal warnings along with spelling errors. There ought to be a way to catch everything at once without running `sphinx-build` twice, but it seems like that would require calling Sphinx from a custom Python script (in my understanding that would imply extending our `conf.py`), that should also eliminate the need for grepping thing out etc. That's only a theory, and I'd rather avoid diving into that for now.

joestringer commented on Documentation/check-build.sh:
```diff
@@ -0,0 +1,81 @@
+}
+
+filter_warnings() {
+    grep -v \
```
I think that the `grep` match here is not quite having the intended effect, because the return code is based on the matched expressions (even if the output for those matches is excluded), it's not based on whether there is other content in the file.

Maybe for the first run, you could just run this grep then pipe the output to `wc -l` and if there's more than zero lines, we need to print them out and fail out.

See the ignored WARNING below about the `admin_guide` link, the build was skipping past that and only complaining about my spelling:

```
$ git diff                                                                                                                                                                     [0/1878]
diff --git a/Documentation/troubleshooting.rst b/Documentation/troubleshooting.rst
index f2e1d29a9947..4f6c896f068f 100644
--- a/Documentation/troubleshooting.rst
+++ b/Documentation/troubleshooting.rst
@@ -6,6 +6,8 @@

 .. _admin_guide:

+@Im'a shootin' for trouble
+
 ###############
 Troubleshooting
 ###############
$ make -C Documentation/
Sending build context to Docker daemon  10.24kB
Step 1/5 : FROM docker.io/library/python:3.7.5-alpine3.10
 ---> 8922d588eec6
Step 2/5 : LABEL maintainer="maintainer@cilium.io"
 ---> Using cache
 ---> be4f82463a09
Step 3/5 : RUN apk add --no-cache --virtual --update     aspell-en     bash     ca-certificates     enchant     git     py-pip     python     sphinx-python     && true
 ---> Using cache
 ---> 3d659eafb699
Step 4/5 : ADD ./requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> ef81ad334356
Step 5/5 : RUN pip install -r /tmp/requirements.txt
 ---> Using cache
 ---> 17243537fa83
Successfully built 17243537fa83
Successfully tagged cilium/docs-builder:latest
  GEN   Documentation/_api
  GEN   Documentation/_build/html
Validating documentation (syntax, spelling)...
/usr/local/lib/python3.7/site-packages/sphinx/application.py:402: RemovedInSphinx20Warning: app.info() is now deprecated. Use sphinx.util.logging instead.
  RemovedInSphinx20Warning)
/src/Documentation/index.rst:32: WARNING: undefined label: admin_guide (if the link has no caption the label must precede a section header)
/usr/local/lib/python3.7/site-packages/sphinx/application.py:388: RemovedInSphinx20Warning: app.warning() is now deprecated. Use sphinx.util.logging instead.
  RemovedInSphinx20Warning)
WARNING: Found 2 misspelled words
WARNING: Not copying tabs assets! Not compatible with spelling builder

Please fix the following spelling mistakes:
* Documentation/troubleshooting.rst:9: (Im'a)
* Documentation/troubleshooting.rst:9: (shootin)

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh Im\\'a shootin
make: *** [Makefile:28: html] Error 1

```

errordeveloper replied:
For some reason the redirect with `2>&1 > warnings.txt` ceased working actually, but simple `2>` solves the issues.

The output looks like this now:

```
$ make html
tar c requirements.txt Dockerfile \
	  | docker build --tag cilium/docs-builder -
Sending build context to Docker daemon  10.24kB
Step 1/6 : FROM docker.io/library/python:3.7.5-alpine3.10
 ---> 8922d588eec6
Step 2/6 : LABEL maintainer="maintainer@cilium.io"
 ---> Using cache
 ---> f341bfd1da19
Step 3/6 : RUN apk add --no-cache --virtual --update     aspell-en     bash     ca-certificates     enchant     git     py-pip     python     sphinx-python     && true
 ---> Using cache
 ---> f2ee248a51ae
Step 4/6 : ADD ./requirements.txt /tmp/requirements.txt
 ---> Using cache
 ---> 9b21a9a640e6
Step 5/6 : RUN pip install -r /tmp/requirements.txt
 ---> Using cache
 ---> a0be0a621785
Step 6/6 : RUN install -m 0777 -d /usr/local/lib/python3.7/site-packages/versionwarning/_static/data
 ---> Using cache
 ---> 93f6ab089dec
Successfully built 93f6ab089dec
Successfully tagged cilium/docs-builder:latest
cp -r ../api _api
docker container run --rm --workdir /src/Documentation --volume /Users/ilya/Code/cilium/Documentation/..:/src --user "501:20" cilium/docs-builder ./check-build.sh html
Validating documentation (syntax, spelling)...

Please fix the following documentation warnings:
/src/Documentation/index.rst:32: WARNING: undefined label: admin_guide (if the link has no caption the label must precede a section header)

Please fix the following spelling mistakes:
* Documentation/troubleshooting.rst:9: (bdabl) 
* Documentation/troubleshooting.rst:9: (dasd) 

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh bdabl dasd 
make: *** [html] Error 1
```

joestringer commented on Documentation/Makefile:
```diff
@@ -1,93 +1,44 @@
+DOCS_PORT = 9081
+
+run-server: stop-server
+	$(QUIET)docker container run --rm \
```
Can we make this run as a regular user UID/GID?

I ran the following commands and hit trouble:

```
$ make -C Documentation/
...
$ make -C Documentation/ clean
rm: cannot remove '_build/doctrees/environment.pickle': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/k8s-install-connectivity-test.doctree': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/kube-router.doctree': Permission denied
rm: cannot remove '_build/doctrees/gettingstarted/k8s-install-azure-cni-steps.doctree': Permission denied
...
$ ls -lh Documentation/_build/doctrees/environment.pickle
-rw-r--r-- 1 root root 704K Mar 13 10:17 Documentation/_build/doctrees/environment.pickle
```

errordeveloper replied:
Good point, I'll see what can be done. I only tested on Docker for Mac, and the [`osxfs`](https://docs.docker.com/docker-for-mac/osxfs/) does remaps UID/GID.

errordeveloper replied:
See https://github.com/cilium/cilium/pull/10572/files#r392967432.

joestringer reviewed (changes requested):
This is looking much simpler, awesome!

I think that there's some basic behavioural differences still between the existing implementation and the new one, and ideally I don't want to reintroduce regressions where the warnings are ignored because we've been there before and things get out of hand quickly ;) I did a little bit of investigation below, hopefully that helps guide getting the new version right here.

Unfortunately I don't think we have any sort of basic testing bootstrap; what I'd ideally want is some very simple patches that deliberately add different types of errors into the docs, then run the docs targets to make sure they fail --- I'm thinking syntax errors, spelling errors, cmdref differences.

On cmdref differences, I didn't investigate this, but I'd want to make sure that if you add a flag to the daemon, the `cmdref` target here correctly generates the new version of the docs, and the docs test build will fail because the developer hasn't added those cmdref changes to the documentation.

errordeveloper committed:
docs: Run builder container as host UID/GID

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper commented on Documentation/Makefile:
```diff
@@ -1,93 +1,45 @@
+DOCKER_RUN := docker container run --rm \
+		--workdir /src/Documentation \
+		--volume $(CURDIR)/..:/src \
+		--user "$(shell id -u):$(shell id -g)" \
```
@joestringer this should work fine on Linux, would you mind to test please?

joestringer replied:
Works for me, thanks!

qmonnet commented on Documentation/.dockerignore:
```diff
@@ -0,0 +1,3 @@
+# Ingore everything except for requirements file
```
Typo: Ingore

qmonnet commented on Documentation/Dockerfile:
```diff
@@ -2,26 +2,16 @@ FROM docker.io/library/python:3.7.5-alpine3.10
-ADD _api $API_DIR
-ADD . $DOCS_DIR
+ADD ./requirements.txt /tmp/requirements.txt
+RUN pip install -r /tmp/requirements.txt
```
GitHub says end-of-line is missing?

errordeveloper replied:
Seems like a standard vscode behaviour.. let me see if there is a setting for it.

qmonnet commented on Documentation/Dockerfile:
```diff
@@ -2,26 +2,16 @@ FROM docker.io/library/python:3.7.5-alpine3.10
     bash \
     ca-certificates \
-    make \
+    enchant \
```
Is this distinct from pyenchant, which should be installed with the requirements?

errordeveloper replied:
Good question... I haven't questioned these dependencies, keeping them as they are for the purpose of this PR :)

qmonnet commented on Documentation/check-build.sh:
```diff
@@ -0,0 +1,69 @@
+    fi
+fi
+
+exec sphinx-build -M "${target}" "${source_dir}" "${build_dir}"
```
No end-of-line?

qmonnet commented on Documentation/update-cmdrefs.sh:
```diff
@@ -0,0 +1,20 @@
+
+for g in "${generators[@]}" ; do
+    "${source_dir}/${g}" "${cmdref_dir}"
+done
```
Nor here?

joestringer reviewed (changes requested):
I found that this broke the build:

```
$ make
...
  GEN   Documentation/cmdref
./update-cmdrefs.sh: line 19: /src/Documentation/../cilium/cilium cmdref -d: No such file or directory
```

The `cmdref` target depends upon the compiled binaries from the tree to generate the documentation.

aanm commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,14 @@
+name: Documentation Updates
+
+on: [push]
+
+jobs:
+    build-html:
+        name: Validate & Build HTML
+        runs-on: ubuntu-18.04
+        steps:
+            - uses: actions/checkout@v1
+            - uses: docker://cilium/docs-builder:latest
+              with:
+                entrypoint: ./Documentation/check-build.sh
+                args: html
```
Will this work if external contributors push a change?

errordeveloper replied:
I don't see a reason why not, GitHub actions that use the API are problematic with external folks, but this is just simple read-only job.

errordeveloper committed:
docs: Ensure builder image is built first

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
docs: Set HOME

Otherwise we get an error where some dependency is trying to write
to $HOME/.config.

$ /src/daemon/cilium-agent --cmdref /src/Documentation/cmdref
unable to start gops: mkdir /.config: permission denied

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

errordeveloper committed:
docs: Few minor improvements

- Change how cmdref checks are implemented, use git diff instead
  of temporary files, move the script into Documentation dir

- Remove SPHINXOPS from Makefile.quite, as it's an implementation
  detail that should be confined to Documentation/Makefile

- make check-build.sh more consisten with other docs build script

Signed-off-by: Ilya Dmitrichenko <errordeveloper@gmail.com>

joestringer commented on Documentation/check-cmdref.sh:
```diff
@@ -0,0 +1,14 @@
+
+if ! git diff --quiet -- "${cmdref_dir}" ; then
+    git --no-pager diff "${cmdref_dir}"
+    echo "HINT: to fix this, run 'make -C Documetation update-cmdref ; git commit Documetation/cmdref --message \"docs: Update cmdref\"'"
```
```suggestion
    echo "HINT: to fix this, run 'make -C Documentation update-cmdref ; git commit Documentation/cmdref --message \"docs: Update cmdref\"'"
```

pchaigno replied:
This check currently fails for me, when running `make -j$(nproc)` on a clean clone. It reports all cmdref files as missing:
```
rm -rvf cmdref/cilium*.md
docker container run --rm --workdir /src/Documentation --volume /home/paul/cilium/Documentation/..:/src --user "1000:1000" cilium/docs-builder ./check-cmdref.sh
removed 'cmdref/cilium-agent.md'
removed 'cmdref/cilium-bugtool.md'
[...]
removed 'cmdref/cilium_version.md'
docker container run --rm --workdir /src/Documentation --volume /home/paul/cilium/Documentation/..:/src --user "1000:1000" cilium/docs-builder ./update-cmdref.sh
diff --git a/Documentation/cmdref/cilium-agent.md b/Documentation/cmdref/cilium-agent.md
deleted file mode 100644
index 2729719e1..000000000
--- a/Documentation/cmdref/cilium-agent.md
+++ /dev/null
@@ -1,174 +0,0 @@
-<!-- This file was autogenerated via cilium-agent --cmdref, do not edit manually-->
-
[...]
diff --git a/Documentation/cmdref/cilium-bugtool.md b/Documentation/cmdref/cilium-bugtool.md
deleted file mode 100644
index acd702818..000000000
--- a/Documentation/cmdref/cilium-bugtool.md
+++ /dev/null
[...]
```

errordeveloper replied:
@pchaigno thanks - well spotted! I am actually not sure how and whether to fix this to be honest, I am not yet equipped to run the entire build locally, and overall the Makefile is pretty complex, this could be to do with anything really. If you posted the full output, I can take a look and see what it maybe. I am not sure if we really must support `-j`, in my experience it break quite often, unless it's specifically be take care of.
Also, `-j` is most helpful when you have to compile a lot of `.c` files in `.o` files and you are invoking the compiler for each one of them, which is something you don't need to worry about with `go build`  as it runs as many gorouties as it sees fit. It's even less useful when you are invoking `docker run`. Basically you need to get dependencies really straight for `-j` to work well, we declare some dependencies, but it's also done a little naively right now. I also have serious doubts about using make dependencies with `go build`, especially since build cache was improved recently.

errordeveloper replied:
Aside from what I said about - does `make -j$(nproc)` work on master? If it did, I will have to fix it here or follow-up with another PR.

pchaigno replied:
Yes, `make -j12` works on master and saves about 32 seconds on a 53 seconds build compared to `make`. I haven't looked into what is actually impacted (maybe BPF compilation?).

I'm sending you the full output of `make -j12 &> build-output.txt` via Slack (GitHub upload seems to be temporarily failing).

errordeveloper replied:
> saves about 32 seconds on a 53 seconds build

That's not for a clean build, right?

pchaigno replied:
I think it is. At least, I ran `make clean` and `git clean -fdx` prior to those test builds.

pchaigno replied:
With this pull request, running `make` takes 10 more seconds on the first run and 10 less seconds once Docker images are prepared. Same impact on `make -j12`. I'm down to a <10s build time! :tada:

errordeveloper replied:
That's a nice side-effect indeed,  but the main goal was to actually make doc issues fail faster with GitHub Actions instead of Jenkins.

errordeveloper replied:
The `make -j` issue should now be resolved per Slack discussion.

joestringer commented on Makefile:
```diff
@@ -477,18 +461,10 @@ install-manpages:
 postcheck: build
-	@$(ECHO_CHECK) contrib/scripts/check-cmdref.sh
-	$(QUIET) MAKE=$(MAKE) contrib/scripts/check-cmdref.sh
+	$(QUIET)$(MAKE) -C Documentation update-cmdref check-cmdref
```
(not related to this specific line, but in this area)

If I currently use `make check-docs` or `make test-docs` for simple docs validation, what do you recommend I do instead? `make -C Documentation`? `make render-docs`? (I guess the latter is generally more popular anyway)

errordeveloper replied:
Yeah `make render-docs` will run the general check logic, but it also creates the Apache container, which is what it used to do also. I think `make -C Documentation` (alias for `make -C Documentation html`) will do the the checks and just write the HTML files, I avoided promoting too many new targets to the top-level makefile for the time being.

I just double-checked, and `make -C Documentation html` certainly covers what `make check-docs` as well as `make test-docs` did.

joestringer reviewed (approved):
I didn't do a deep review this time since I did that before, but I noticed some minor glancing nits.

Looks like this needs a rebase. With that and the typo below fixed, I think it should be good to go. Nice work!

pchaigno commented on Documentation/Dockerfile:
```diff
@@ -2,26 +2,17 @@ FROM docker.io/library/python:3.7.5-alpine3.10
-RUN apk add --no-cache --virtual --update $PACKAGES && \
-    pip install --upgrade pip && \
-    pip install -r $DOCS_DIR/requirements.txt
+    && true
```
Why continue if that command fails? Won't subsequent commands fail as well?

errordeveloper replied:
This is just a stylistic/convenience thing, it allows you to sort package alphabetically without worrying about `\`, it doesn't do anything actually. E.g. `false && true` is gonna be `false`.

pchaigno replied:
Oh, I see. I thought this was the same trick one needs to use with `set -e`. Thanks for the explanation!

pchaigno commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,14 @@
+name: Documentation Updates
+
+on: [push]
```
Will GitHub trigger this action if someone develops a Documentation change in a fork then opens a pull request? I.e., do we need to add `pull_request` here?

errordeveloper replied:
I believe it will, the `pull_request` event in my understanding is specially about PRs as such, you can write a bot that comments on certain PRs, for example. Push is a more general one. We can test this once it's merged to be sure :)
